File: .repo/system-design-mastery-module-2-database-fundamentals/1-indexing-and-query-optimization/1-java/src/main/java/academy/starci/indexing/products/ProductsService.java
package academy.starci.indexing.products;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.*;

@Service
public class ProductsService {

    private static final String[] CATEGORIES = {
        "Electronics", "Clothing", "Books", "Home & Garden", "Sports", "Toys", "Food & Beverage", "Health", "Automotive", "Music"
    };

    private static final String[] BRANDS = {
        "Acme", "Zenith", "Pinnacle", "Vertex", "Nova", "Apex", "Stellar", "Orbit", "Quantum", "Flux", "Titan", "Vortex", "Prism", "Nexus", "Echo"
    };

    private static final String[] ADJECTIVES = {
        "Premium", "Ultra", "Pro", "Elite", "Classic", "Advanced", "Essential", "Deluxe", "Standard", "Compact"
    };

    private static final String[] NOUNS = {
        "Widget", "Gadget", "Device", "Tool", "Kit", "System", "Module", "Unit", "Pack", "Set"
    };

    private final JdbcTemplate jdbcTemplate;

    public ProductsService(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    public Map<String, Object> seed(int count) {
        Random rand = new Random();
        int batchSize = 500;
        int inserted = 0;

        String sql = "INSERT INTO products (id, name, sku, price, category, brand, description, stock, rating, created_at) " +
                     "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, NOW())";

        for (int i = 0; i < count; i += batchSize) {
            int currentBatchSize = Math.min(batchSize, count - i);
            List<Object[]> batchArgs = new ArrayList<>();

            for (int j = 0; j < currentBatchSize; j++) {
                String adjective = ADJECTIVES[rand.nextInt(ADJECTIVES.length)];
                String noun = NOUNS[rand.nextInt(NOUNS.length)];
                String brand = BRANDS[rand.nextInt(BRANDS.length)];
                String name = brand + " " + adjective + " " + noun;
                String sku = String.format("%c%c-%d-%d", 
                    (char)('A' + rand.nextInt(26)), 
                    (char)('A' + rand.nextInt(26)), 
                    System.currentTimeMillis(), 
                    i + j);
                double price = Math.round((rand.nextDouble() * 999.0 + 1.0) * 100.0) / 100.0;
                String category = CATEGORIES[rand.nextInt(CATEGORIES.length)];
                String description = String.format("High-quality %s %s by %s. Built for performance and reliability.", 
                    adjective.toLowerCase(), noun.toLowerCase(), brand);
                int stock = rand.nextInt(1000);
                double rating = Math.round((rand.nextDouble() * 4.0 + 1.0) * 10.0) / 10.0;

                batchArgs.add(new Object[]{
                    UUID.randomUUID(),
                    name,
                    sku,
                    price,
                    category,
                    brand,
                    description,
                    stock,
                    rating
                });
            }

            jdbcTemplate.batchUpdate(sql, batchArgs);
            inserted += currentBatchSize;
        }

        Map<String, Object> result = new HashMap<>();
        result.put("inserted", inserted);
        return result;
    }

    public Map<String, Object> findWithoutIndex(String category) {
        long start = System.nanoTime();
        List<Map<String, Object>> rows = jdbcTemplate.execute((Connection conn) -> {
            try (var stmt = conn.createStatement()) {
                stmt.execute("SET enable_indexscan = OFF");
                stmt.execute("SET enable_bitmapscan = OFF");

                List<Map<String, Object>> results = new ArrayList<>();
                String query = "SELECT * FROM products WHERE category = ? ORDER BY price ASC";
                try (PreparedStatement pStmt = conn.prepareStatement(query)) {
                    pStmt.setString(1, category);
                    try (ResultSet rs = pStmt.executeQuery()) {
                        while (rs.next()) {
                            Map<String, Object> row = new LinkedHashMap<>();
                            row.put("id", rs.getString("id"));
                            row.put("name", rs.getString("name"));
                            row.put("sku", rs.getString("sku"));
                            row.put("price", rs.getString("price"));
                            row.put("category", rs.getString("category"));
                            row.put("brand", rs.getString("brand"));
                            row.put("description", rs.getString("description"));
                            row.put("stock", rs.getInt("stock"));
                            row.put("rating", rs.getDouble("rating"));
                            var ts = rs.getTimestamp("created_at");
                            row.put("createdAt", ts != null ? ts.toInstant().toString() : null);
                            results.add(row);
                        }
                    }
                }

                stmt.execute("SET enable_indexscan = ON");
                stmt.execute("SET enable_bitmapscan = ON");
                return results;
            }
        });

        double duration = (System.nanoTime() - start) / 1_000_000.0;

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("rows", rows);
        result.put("executionTimeMs", Math.round(duration * 1000.0) / 1000.0);
        return result;
    }

    public Map<String, Object> findWithIndex(String category, double minPrice) {
        long start = System.nanoTime();
        String query = "SELECT * FROM products WHERE category = ? AND price >= ? ORDER BY price ASC";
        List<Map<String, Object>> rows = jdbcTemplate.query(query, (rs, rowNum) -> {
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("id", rs.getString("id"));
            row.put("name", rs.getString("name"));
            row.put("sku", rs.getString("sku"));
            row.put("price", rs.getString("price"));
            row.put("category", rs.getString("category"));
            row.put("brand", rs.getString("brand"));
            row.put("description", rs.getString("description"));
            row.put("stock", rs.getInt("stock"));
            row.put("rating", rs.getDouble("rating"));
            var ts = rs.getTimestamp("created_at");
            row.put("createdAt", ts != null ? ts.toInstant().toString() : null);
            return row;
        }, category, minPrice);

        double duration = (System.nanoTime() - start) / 1_000_000.0;

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("rows", rows);
        result.put("executionTimeMs", Math.round(duration * 1000.0) / 1000.0);
        return result;
    }

    public Map<String, Object> explainQuery(String sql) {
        String planSql = "EXPLAIN ANALYZE " + sql;
        List<String> planLines = jdbcTemplate.query(planSql, (rs, rowNum) -> rs.getString(1));
        String plan = String.join("\n", planLines);
        Map<String, Object> res = new HashMap<>();
        res.put("plan", plan);
        return res;
    }

    public Map<String, Object> compareQueries() {
        Random rand = new Random();
        String category = CATEGORIES[rand.nextInt(CATEGORIES.length)];
        double minPrice = Math.round((rand.nextDouble() * 200.0 + 50.0) * 100.0) / 100.0;

        String sql = String.format(Locale.US, "SELECT * FROM products WHERE category = '%s' AND price >= %.2f ORDER BY price ASC", category, minPrice);

        // Without index
        String planWithout = jdbcTemplate.execute((Connection conn) -> {
            try (var stmt = conn.createStatement()) {
                stmt.execute("SET enable_indexscan = OFF");
                stmt.execute("SET enable_bitmapscan = OFF");

                List<String> lines = new ArrayList<>();
                try (ResultSet rs = stmt.executeQuery("EXPLAIN ANALYZE " + sql)) {
                    while (rs.next()) {
                        lines.add(rs.getString(1));
                    }
                }

                stmt.execute("SET enable_indexscan = ON");
                stmt.execute("SET enable_bitmapscan = ON");
                return String.join("\n", lines);
            }
        });

        // With index
        List<String> linesWith = jdbcTemplate.query("EXPLAIN ANALYZE " + sql, (rs, rowNum) -> rs.getString(1));
        String planWith = String.join("\n", linesWith);

        String timeWithout = extractExecutionTime(planWithout);
        String timeWith = extractExecutionTime(planWith);

        Map<String, Object> res = new LinkedHashMap<>();
        
        Map<String, Object> withoutIdx = new LinkedHashMap<>();
        withoutIdx.put("plan", planWithout);
        withoutIdx.put("executionTimeMs", timeWithout);

        Map<String, Object> withIdx = new LinkedHashMap<>();
        withIdx.put("plan", planWith);
        withIdx.put("executionTimeMs", timeWith);

        res.put("withoutIndex", withoutIdx);
        res.put("withIndex", withIdx);
        res.put("summary", String.format(Locale.US, "Query: SELECT * FROM products WHERE category='%s' AND price >= %.2f ORDER BY price ASC", category, minPrice));

        return res;
    }

    private String extractExecutionTime(String plan) {
        java.util.regex.Matcher m = java.util.regex.Pattern.compile("Execution Time:\\s+([\\d.]+)\\s+ms").matcher(plan);
        if (m.find()) {
            return m.group(1) + " ms";
        }
        return "N/A";
    }

    public Map<String, Object> getTableStats() {
        String statsSql = "SELECT relname AS table_name, seq_scan, idx_scan, n_live_tup AS live_rows " +
                           "FROM pg_stat_user_tables WHERE relname = 'products'";

        List<Map<String, Object>> statsList = jdbcTemplate.query(statsSql, (rs, rowNum) -> {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("table_name", rs.getString("table_name"));
            map.put("seq_scan", rs.getLong("seq_scan"));
            map.put("idx_scan", rs.getLong("idx_scan"));
            map.put("live_rows", rs.getLong("live_rows"));
            return map;
        });

        Map<String, Object> tableStats = statsList.isEmpty() ? new HashMap<>() : statsList.get(0);

        String indexSql = "SELECT indexrelname AS index_name, idx_scan AS times_used, " +
                          "pg_size_pretty(pg_relation_size(indexrelid)) AS index_size " +
                          "FROM pg_stat_user_indexes WHERE relname = 'products'";

        List<Map<String, Object>> indexStats = jdbcTemplate.query(indexSql, (rs, rowNum) -> {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("index_name", rs.getString("index_name"));
            map.put("times_used", rs.getLong("times_used"));
            map.put("index_size", rs.getString("index_size"));
            return map;
        });

        Map<String, Object> res = new LinkedHashMap<>();
        res.put("tableStats", tableStats);
        res.put("indexStats", indexStats);
        return res;
    }
}

File: .repo/system-design-mastery-module-2-database-fundamentals/1-indexing-and-query-optimization/0-typescript/src/products/products.service.ts
/**
 * Products service — seed data, query with/without index, explain plan, compare.
 */
import {
    Injectable,
    Logger,
} from "@nestjs/common"
import {
    InjectRepository,
} from "@nestjs/typeorm"
import {
    DataSource,
    Repository,
} from "typeorm"
import {
    ProductEntity,
} from "./entities/product.entity"

/**
 * Sample product categories.
 */
const CATEGORIES = [
    "Electronics",
    "Clothing",
    "Books",
    "Home & Garden",
    "Sports",
    "Toys",
    "Food & Beverage",
    "Health",
    "Automotive",
    "Music",
]

/**
 * Sample brands.
 */
const BRANDS = [
    "Acme",
    "Zenith",
    "Pinnacle",
    "Vertex",
    "Nova",
    "Apex",
    "Stellar",
    "Orbit",
    "Quantum",
    "Flux",
    "Titan",
    "Vortex",
    "Prism",
    "Nexus",
    "Echo",
]

/**
 * Sample adjectives for product names.
 */
const ADJECTIVES = [
    "Premium",
    "Ultra",
    "Pro",
    "Elite",
    "Classic",
    "Advanced",
    "Essential",
    "Deluxe",
    "Standard",
    "Compact",
]

/**
 * Sample nouns for product names.
 */
const NOUNS = [
    "Widget",
    "Gadget",
    "Device",
    "Tool",
    "Kit",
    "System",
    "Module",
    "Unit",
    "Pack",
    "Set",
]

@Injectable()
export class ProductsService {
    private readonly logger = new Logger(ProductsService.name)

    constructor(
        @InjectRepository(ProductEntity)
        private readonly productRepository: Repository<ProductEntity>,
        private readonly dataSource: DataSource,
    ) {}

    /**
     * Pick a random element from an array.
     */
    private randomElement<T>(arr: T[]): T {
        return arr[Math.floor(Math.random() * arr.length)] as T
    }

    /**
     * Generate a unique SKU code.
     */
    private generateSku(index: number): string {
        const prefix = String.fromCharCode(65 + Math.floor(Math.random() * 26))
        const suffix = String.fromCharCode(65 + Math.floor(Math.random() * 26))
        return `${prefix}${suffix}-${Date.now()}-${index}`
    }

    /**
     * Seed product data into table — insert in batches of 500.
     */
    async seed(count: number): Promise<{ inserted: number }> {
        this.logger.log(`Seeding ${count} products...`)

        const batchSize = 500
        let inserted = 0

        for (let i = 0; i < count; i += batchSize) {
            const batch: Partial<ProductEntity>[] = []
            const currentBatch = Math.min(batchSize, count - i)

            for (let j = 0; j < currentBatch; j++) {
                const adjective = this.randomElement(ADJECTIVES)
                const noun = this.randomElement(NOUNS)
                const brand = this.randomElement(BRANDS)

                batch.push({
                    name: `${brand} ${adjective} ${noun}`,
                    sku: this.generateSku(i + j),
                    price: parseFloat((Math.random() * 999 + 1).toFixed(2)),
                    category: this.randomElement(CATEGORIES),
                    brand,
                    description: `High-quality ${adjective.toLowerCase()} ${noun.toLowerCase()} by ${brand}. Built for performance and reliability.`,
                    stock: Math.floor(Math.random() * 1000),
                    rating: parseFloat((Math.random() * 4 + 1).toFixed(1)),
                })
            }

            await this.productRepository
                .createQueryBuilder()
                .insert()
                .into(ProductEntity)
                .values(batch)
                .execute()

            inserted += currentBatch
        }

        this.logger.log(`Seeded ${inserted} products`)
        return { inserted }
    }

    /**
     * Query WITHOUT index — disable index scan to force sequential scan.
     */
    async findWithoutIndex(category: string): Promise<{
        rows: ProductEntity[]
        executionTimeMs: number
    }> {
        const start = performance.now()

                // Create separate QueryRunner to disable index scan for this session.
        const queryRunner = this.dataSource.createQueryRunner()
        await queryRunner.connect()

        try {
            await queryRunner.query("SET enable_indexscan = OFF")
            await queryRunner.query("SET enable_bitmapscan = OFF")

            const rows = await queryRunner.query(
                "SELECT * FROM products WHERE category = $1 ORDER BY price ASC",
                [category],
            )

            const executionTimeMs = parseFloat(
                (performance.now() - start).toFixed(3),
            )

            return { rows, executionTimeMs }
        } finally {
                        // Re-enable index scan.
            await queryRunner.query("SET enable_indexscan = ON")
            await queryRunner.query("SET enable_bitmapscan = ON")
            await queryRunner.release()
        }
    }

    /**
     * Query WITH index — leverages composite index (category, price).
     */
    async findWithIndex(
        category: string,
        minPrice: number,
    ): Promise<{
        rows: ProductEntity[]
        executionTimeMs: number
    }> {
        const start = performance.now()

                // Query using composite index (category, price).
        const rows = await this.productRepository
            .createQueryBuilder("p")
            .where("p.category = :category", { category })
            .andWhere("p.price >= :minPrice", { minPrice })
            .orderBy("p.price", "ASC")
            .getMany()

        const executionTimeMs = parseFloat((performance.now() - start).toFixed(3))

        return { rows, executionTimeMs }
    }

    /**
     * Run EXPLAIN ANALYZE on arbitrary SQL — view execution plan.
     */
    async explainQuery(sql: string): Promise<{ plan: string }> {
        const result = await this.dataSource.query(`EXPLAIN ANALYZE ${sql}`)
        const plan = result.map((row: Record<string, string>) => row["QUERY PLAN"]).join("\n")
        return { plan }
    }

    /**
     * Compare same query with index vs without index — return plan + timing.
     */
    async compareQueries(): Promise<{
        withoutIndex: { plan: string; executionTimeMs: string }
        withIndex: { plan: string; executionTimeMs: string }
        summary: string
    }> {
        const category = this.randomElement(CATEGORIES)
        const minPrice = parseFloat((Math.random() * 200 + 50).toFixed(2))

                // Query 1: force sequential scan (disable index).
        const queryRunner = this.dataSource.createQueryRunner()
        await queryRunner.connect()

        let planWithout: string
        let timeWithout: string

        try {
            await queryRunner.query("SET enable_indexscan = OFF")
            await queryRunner.query("SET enable_bitmapscan = OFF")

            const result1 = await queryRunner.query(
                `EXPLAIN ANALYZE SELECT * FROM products WHERE category = '${category}' AND price >= ${minPrice} ORDER BY price ASC`,
            )
            planWithout = result1.map((r: Record<string, string>) => r["QUERY PLAN"]).join("\n")

            const timingMatch = planWithout.match(
                /Execution Time:\s+([\d.]+)\s+ms/,
            )
            timeWithout = timingMatch ? `${timingMatch[1]} ms` : "N/A"
        } finally {
            await queryRunner.query("SET enable_indexscan = ON")
            await queryRunner.query("SET enable_bitmapscan = ON")
            await queryRunner.release()
        }

                // Query 2: use index (default).
        const result2 = await this.dataSource.query(
            `EXPLAIN ANALYZE SELECT * FROM products WHERE category = '${category}' AND price >= ${minPrice} ORDER BY price ASC`,
        )
        const planWith = result2.map((r: Record<string, string>) => r["QUERY PLAN"]).join("\n")

        const timingMatch2 = planWith.match(/Execution Time:\s+([\d.]+)\s+ms/)
        const timeWith = timingMatch2 ? `${timingMatch2[1]} ms` : "N/A"

        return {
            withoutIndex: { plan: planWithout, executionTimeMs: timeWithout },
            withIndex: { plan: planWith, executionTimeMs: timeWith },
            summary: `Query: SELECT * FROM products WHERE category='${category}' AND price >= ${minPrice} ORDER BY price ASC`,
        }
    }

    /**
     * Get products table stats — seq_scan, idx_scan, index sizes.
     */
    async getTableStats(): Promise<{
        tableStats: Record<string, unknown>
        indexStats: Record<string, unknown>[]
    }> {
        const stats = await this.dataSource.query(`
            SELECT
                relname AS table_name,
                seq_scan,
                seq_tup_read,
                idx_scan,
                idx_tup_fetch,
                n_tup_ins AS rows_inserted,
                n_tup_upd AS rows_updated,
                n_tup_del AS rows_deleted,
                n_live_tup AS live_rows,
                n_dead_tup AS dead_rows,
                last_vacuum,
                last_analyze
            FROM pg_stat_user_tables
            WHERE relname = 'products'
        `)

        const indexStats = await this.dataSource.query(`
            SELECT
                indexrelname AS index_name,
                idx_scan AS times_used,
                idx_tup_read AS tuples_read,
                idx_tup_fetch AS tuples_fetched,
                pg_size_pretty(pg_relation_size(indexrelid)) AS index_size
            FROM pg_stat_user_indexes
            WHERE relname = 'products'
        `)

        return {
            tableStats: stats[0] ?? {},
            indexStats,
        }
    }
}
